### DISCUSSION TITLE
Select text and see font name

### DISCUSSION THREAD

**Discussion by robertf** [sumatrapdfreader/sumatrapdf#4230] [Ideas]
It would be useful to be able to select a character and see what font is used for it.

Someone requested this some time ago (https://forum.sumatrapdfreader.org/t/select-text-and-see-font-name-in-tooltip/377). It was asked whether any other PDF reader does this. I'm not aware of one, but it would be a nice feature if it's easy to implement. My workaround now is to open the PDF file in LibreOffice Draw.

Thank you for your great PDF reader :-)

- Robert

**Comment by GitHubRulesOK** [comment]
generally its not simple your best bet is an editor where such data is exposed and Tracker XPDFedit is one of two I use for such initial look BUT beware the font name shown is not necessarily the font that was used for building the PDF as often it may simply say Arial when the encoded data is different. Font names are not very useful.

Here I am playing with new kid on the Block MS replacement for Arial (Aptos) and what does the PDF use is simply CID same as the other font so CID1 and CID2 means nothing whatsoever in my mind except they are PDF names.
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/93cf18f4-8c32-4e86-83c5-305d6d2b352f)

**Reply by robertf** [reply]
Thanks for the information.

I read that the CID (https://en.wikipedia.org/wiki/PostScript_fonts#CID) problem is due to a flaw in whatever software exported the PDF file. I don't really understand why such a flaw should exist.

- Robert

**Reply by GitHubRulesOK** [reply]
@robertf 
Not a Flaw it is how Adobe designed PDF version 1  before Unicode support in version 2 (not well used yet)

inside a PDF we think there are letters but they are not all like that. They are encoded bytes of 1 or 2 byte words per letter
soo A can be (00000001) and B (00000010) or <0000000000000010> or <00000000000000000000000000000010> 
so PDF as a format is humongous waste of memory, to reduce that the double long codes can be simplified by using
> character identifier (CID)  

where `B` is stored as say (B) so only 3 characters per letter same with those longer ones they can be stored as numbers like <0002> and be easier to read However the plaintext character for (B) can also be stored as (2) and work just as efficiently because it is not the text but the numeric value that is used to display the graphics.

That is done by a CMAP so the 8bit word `2` can be mapped to `B` as a graphic which may be 100 bytes per time it is used !!

A CMAP will look like
<02><0002> where 2 is the UNICODE character in a font MAP from 
<0002> to <0042> (U+0042 is the unicode hex value of the character Latin Capital Letter B.)

Overall it is amazing we see anything human after all that and when it goes wrong we say it is Mojibake because was previously more common with Japanese !

The key point is there are no human words in a PDF just strings of numbers that at times look like single letters in a row.

see this Magical PDF has  text that are magic carpets to other realms and Abracadabra looks like this

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/5b17e6ff-880a-43d9-b813-b73be5c9f92b)

```
(9)' 27 20 Td (letters.)'
-27 0 Td(A)' 25 20 Td(a)' 17 20 Td(a)' 18 20 Td(a)' 23 20 Td(a)'
-72 20 Td(b)' 58 20 Td(b)'
-35 20 Td(c)'
17 20 Td(d)'
26.5 20 Td(r)' -58 20 Td(r)'
```
so now you know about CIDs you can tell I wrote it as `(A)(a)(a)(a)(a)(b)(b)(c)(d)(r)(r)` so its Human alphabetic as `Abracadabra` is naturally human order and looks like a human word.

You may then see why to find the whole word I need to enter `A
a a a
a`

**Comment by kazerniel** [comment]
[PDF-XChange Viewer](https://www.pdf-xchange.com/product/downloads/discontinued) has this feature, though the program has been discontinued since 2018.

**Reply by GitHubRulesOK** [reply]
PDFxchange Viewer like Acrobat is the smaller editor of a more commercial one but the difference is that free products don't produce revenue. Adobe don't show more than Font Names as a List (Same as SumatraPDF can) in their "Free DC editing Reader"

So MuPDF in theory can select a character and show its description (Mutool.exe is able to do it command line for a whole file) but as a GUI viewer that would be way beyond its commercial value. You can pay for the commercial versions and write your own functions, but it would be larger and slower to do that for every file format.

at about $60 Tracker PDFXedit is exceptionally good value compared to most others.

**Reply by kazerniel** [reply]
> Adobe don't show more than Font Names as a List (Same as SumatraPDF can) in their "Free DC editing Reader"

Yeah, but the free version of PDF-XChange Viewer does show the font types and sizes included in any selected piece of text, so it works for this purpose 🤷 (I've been using it for years, I didn't even realise it was discontinued until I went looking for its link.)
